pkg/gce-cloud-provider/compute/gce-compute.go

```diff
@@ -593,6 +594,65 @@ func (cloud *CloudProvider) CreateSnapshot(ctx context.Context, volKey *meta.Key
 	}
 }
 
+func (cloud *CloudProvider) ResizeDisk(ctx context.Context, volKey *meta.Key, requestBytes int64) (int64, error) {
+	cloudDisk, err := cloud.GetDisk(ctx, volKey)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get disk: %v", err)
+	}
+
+	sizeGb := cloudDisk.GetSizeGb()
+	requestGb := common.BytesToGb(requestBytes)
+
+	// If disk is already of size equal or greater than requested size, we simply return
+	if sizeGb >= requestGb {
+		return requestBytes, nil
+	}
+
+	switch volKey.Type() {
+	case meta.Zonal:
+		return cloud.resizeZonalDisk(ctx, volKey, requestGb)
+	case meta.Regional:
+		return cloud.resizeRegionalDisk(ctx, volKey, requestGb)
+	default:
+		return -1, fmt.Errorf("could not resize disk, key was neither zonal nor regional, instead got: %v", volKey.String())
+	}
+}
+
+func (cloud *CloudProvider) resizeZonalDisk(ctx context.Context, volKey *meta.Key, requestGb int64) (int64, error) {
+	resizeReq := &compute.DisksResizeRequest{
+		SizeGb: requestGb,
+	}
+	op, err := cloud.service.Disks.Resize(cloud.project, volKey.Zone, volKey.Name, resizeReq).Context(ctx).Do()
+	if err != nil {
+		return -1, fmt.Errorf("failed to resize zonal volume %v: %v", volKey.String(), err)
+	}
+
+	err = cloud.waitForZonalOp(ctx, op, volKey.Zone)
+	if err != nil {
+		return -1, fmt.Errorf("failed waiting for op for zonal resize for %s: %v", volKey.String(), err)
+	}
+
+	return requestGb, nil
```

Comment:
do we want to validate the size after success?

Reply:
The `ControllerExpandVolume` layer could do a `GetDisk` and check the size. Same question on the node side, if the GCE backend returns success do we really need to double check? I don't think we do this for any of the other calls

Reply:
For controller resize, I don't think it's as necessary since we pass in the size to the operation